design-proposals/klip-52-bytes-data-type-support.md

```diff
@@ -0,0 +1,142 @@
+key in a bytes schema, but this will not be included in this KLIP.
+
+## Public APIS
+
+### BYTES
+
+The BYTES data type will store an array of raw bytes of an unspecified length. The maximum size of
+the array is limited by the maximum size of a Kafka message, as well as possibly by the value format being used.
+The syntax is as follows:
+
+```roomsql
+CREATE STREAM stream_name (b BYTES, COL2 STRING) AS ...
+CREATE TABLE table_name (col1 STRUCT<field BYTES>) AS ...
+```
+
+By default, BYTES will be displayed in console as HEX strings, where each byte is represented by two characters.
+For example, the byte array `[91, 67]` will be displayed as:
+
+```roomsql
+> SELECT b from STREAM;
+'0x5B43'
+```
+
+Implicit conversions to BYTES will not be supported.
+
+### UDF
+
+The following UDFs will be added:
+
+* `to_bytes(string, encoding)` - this will convert a STRING value to BYTES in the specified encoding.
+The accepted encoders are `hex`, `utf8`, `ascii`, and `base64`.
+* `from_bytes(bytes, encoding)` - this will convert a BYTES value to STRING in the specified encoding.
+The accepted encoders are `hex`, `utf8`, `ascii`, and `base64`.
+
+We will also update some of the existing STRING functions to accept BYTES as a parameter. In general, if a function works on ASCII characters for a STRING parameter,
+then it will work on bytes for a BYTES parameter.
+
+* `len(bytes)` - This will return the length of the stored ByteArray.
+* `concat(bytes...)` - Concatenate an arbitrary number of byte fields
+* `r/lpad(bytes, target_length, padding_bytes)` - pads input BYTES beginning from the left/right with the specified padding BYTES until the target length is reached.
+* `replace(bytes, old_bytes, new_bytes)` - returns the given BYTES value with all occurrences of `old_bytes` with `new_bytes`
+* `split(bytes, delimiter)` - splits a BYTES value into an array of BYTES based on a delimiter
+* `splittomap(bytes, entryDelimiter, kvDelimiter)` - splits a BYTES value into key-value pairs based on a delimiter and creates a MAP from them
+* `substring(bytes, to, from)` - returns the section of the BYTES from the byte at position `to` to `from`
+
+## Design
+### Serialization/Deserialization
+
+BYTES will be handled by [`java.nio.ByteBuffer`](https://docs.oracle.com/javase/7/docs/api/java/nio/ByteBuffer.html) within ksqlDB.
+The underlying Kafka Connect type is the primitive `bytes` type. 
+
+#### Avro
+
+`bytes` is a primitive Avro type. When converting to/from Connect data, the Avro converter ksqlDB
+uses converts byte arrays to ByteBuffer.
+
+#### Protobuf
+
+`bytes` is a primitive Protobuf type. The maximum number of bytes in a byte array is 2<sup>32</sup>.
+When converting to/from Connect data, the Avro converter ksqlDB uses converts byte arrays to ByteBuffer.
+
+#### JSON/Delimited
+
+Byte arrays will be stored in JSON and CSV files as [Base64 MIME](https://docs.oracle.com/javase/8/docs/api/java/util/Base64.html#mime) encoded binary values.
```

Comment:
Is base64 how connect also serializes bytes types?

I wonder if we should have (not on this PR) an encoding method as part of the column type, i.e. `create stream s1 (b bytes encoding 'hex')`. It might even work with strings when we want them utf-16 or utf-8. Just throwing ideas here.

Reply:
For JSON yes. For Delimited, it's configurable, but the default is UTF_8. I'd still rather have them be the same though.

Yeah I think that would be nice to have if the all the SR serializer encodings become configurable.

Reply:
We could make our own serializers for this. Anyway, it is a nice suggestion for the future.